### ISSUE TITLE
[eslint-plugin-react-hooks] allow configuring custom hooks as "stable"

### ISSUE THREAD

**Issue by grncdr** [facebook/react#16873]
**Do you want to request a *feature* or report a *bug*?**

Feature/enhancement

**What is the current behavior?**

Currently the eslint plugin is unable to understand when the return value of a custom hook is "stable".

Example:

```jsx
import React from 'react'

function useToggle(init = false) {
  const [state, setState] = React.useState(init)
  return [state, React.useCallback(() => { setState(v => !v) }]
}

function MyComponent({someProp}) {
  const [enabled, toggleEnabled] = useToggle()

  const handler = React.useCallback(() => {
    toggleEnabled()
    doSomethingWithTheProp(someProp)
  }, [someProp]) // exhaustive-deps warning for toggleEnabled

  return <button onClick={handler}>Do something</button>
}
```

**What is the expected behavior?**

I would like to configure `eslint-plugin-react-hooks` to tell it that `toggleEnabled` is stable and doesn't need to be included in a dependency array. This isn't a huge deal but more of an ergonomic papercut that discourages writing/using custom hooks.

As for how/where to configure it, I would be happy to add something like this to my .eslintrc:

```js
{
  "stableHooks": {
    "useToggle": [false, true],  // first return value is not stable, second is
    "useForm": true,             // entire return value is stable 
  }
}
```

Then the plugin could have an additional check [after these 2 checks](https://github.com/facebook/react/blob/8b580a89d6dbbde8a3ed69475899addef1751116/packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js#L228-L231) that tests for custom names.

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

All versions of eslint-plugin-react-hooks have the same deficiency.

**Comment by grncdr** [comment]
I went ahead and implemented this (see above commit) just to see how it would play out in my own codebase. If anybody else feels like trying it, I've published it to npm under `@grncdr/eslint-plugin-react-hooks`. You will need to update your eslintrc to reference the scoped plugin name and configure your static hook names:

```patch
-  "plugins": ["react-hooks"],
+  "plugins": ["@grncdr/react-hooks"],
   "rules": {
-    "react-hooks/rules-of-hooks": "error",
-    "react-hooks/exhaustive-deps": "warn",
+    "@grncdr/react-hooks/rules-of-hooks": "error",
+    "@grncdr/react-hooks/exhaustive-deps": [
+      "error",
+      {
+        "additionalHooks": "usePromise",
+        "staticHooks": {
+          "useForm": true,
+          "useEntityCache": true,
+          "useItem": [false, true],
+          "useQueryParam": [false, true]
+        }
+      }
+    ],
```

If anybody from the React team thinks the idea is worth pursuing I'll try to add some tests and make a proper PR.

**Comment by VanTanev** [comment]
This seems like a really great addition, would love to see it in react-hooks

**Comment by grncdr** [comment]
@VanTanev have you tried my fork? I've been using it since my last comment and haven't had any issues, but positive experience from others would presumably be interesting to the maintainers.

**Comment by ramiel** [comment]
Any news on this. It's very annoying now because you cannot use reliably this lint rule when you use custom hook, so you have to disable the rule leading to potential dangerous situations

**Comment by kravets-levko** [comment]
IMHO it would be great if this rule could detect some common "static" patterns in custom hook, for example if custom hook returns result of `useRef()`/`useCallback(..., [])`/`useMemo(..., [])` etc.

**Comment by ramiel** [comment]
Indeed. Still there may be ambiguous situations  and so having the ability to set it up through options could still be needed

**Comment by garetmckinley** [comment]
Commenting to bump this thread and show my interest. Working on a large codebase with lots of custom hooks means that this would allow us to more reliably use the hooks linter. I understand that the reason they might not want to allow this, is because it could enable people to introduce dangerous bugs into their apps. So maybe it's a feature that should be added with a large disclaimer.

It's pretty likely that the maintainers simply don't want to deal with bug reports that are related to people setting their hooks to static when they actually aren't static. A lot of people will misunderstand what it means to have static hooks.

**Comment by grncdr** [comment]
> IMHO it would be great if this plugin could detect some common "static" patterns in custom hook, for example if custom hook returns result of `useRef()`/`useCallback(..., [])`/`useMemo(..., [])` etc.

This is way beyond the scope of what ESLint can do (it would require whole-program taint-tracking) so definitely not going to happen here.

> I understand that the reason they might not want to allow this, is because it could enable people to introduce dangerous bugs into their apps. So maybe it's a feature that should be added with a large disclaimer.
>
> It's pretty likely that the maintainers simply don't want to deal with bug reports that are related to people setting their hooks to static when they actually aren't static. A lot of people will misunderstand what it means to have static hooks.

I would totally understand this point of view, but until somebody from the React team replies, I'll keep hoping (and using my fork 😉).

**Comment by ksjogo** [comment]
@grncdr can you please point me to the source of your folk?

**Comment by grncdr** [comment]
@ksjogo sure, my changes are in this branch: https://github.com/grncdr/react/tree/eslint-plugin-react-hooks-static-hooks-config

**Comment by douglasjunior** [comment]
This is really missing for us, because we have hooks like `useAxios` that always return the same value.

We have faced problems such as:

```js
const axios = useAxios(...);

const requestSomething = useCallback(() => {
      return axios.get(...);
}, []);
```
ESLint warning:
```
React Hook useCallback has a missing dependency: 'axios'. Either include it or remove the dependency array.eslint(react-hooks/exhaustive-deps)
```

**Comment by grncdr** [comment]
I’m curious about that use case: what is the useAxios hook doing that couldn’t be accomplished with a normal import?

**Comment by douglasjunior** [comment]
> I’m curious about that use case: what is the useAxios hook doing that couldn’t be accomplished with a normal import?

Internally it uses `useMemo` to create an instance of axios, and also a `useEffect` that cancels pending requests when the component is unmounted.

Additionally, it configures the `baseUrl` and automatically injects the authentication token via interceptor.

**Comment by acnebs** [comment]
I would also like to see this behavior, mostly just for `setState` and `useRef`.

@douglasjunior don't want to get too off-topic, but you might just wanna have a global/singleton/etc. for that? Seems unnecessary to set the `baseUrl` and token every time you use the hook, as presumably those values will not change between instances of the hook.

**Comment by douglasjunior** [comment]
> @douglasjunior don't want to get too off-topic, but you might just wanna have a global/singleton/etc. for that? Seems unnecessary to set the `baseUrl` and token every time you use the hook, as presumably those values will not change between instances of the hook.

The `useAxios` is configurable, it can receive a custom baseURL, and others configs. 

But in the end it makes no difference, the main purpose for us is to cancel pending requests, and make the axios instance private to the component.

**Comment by squirly** [comment]
Allowing configuration of the dependency argument position would be useful as well.

It is currently hard coded to `0` for `additionalHooks`:
https://github.com/facebook/react/blob/8b580a89d6dbbde8a3ed69475899addef1751116/packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js#L1361

This allows support for hooks that take more than 2 arguments. Eg.:
```js
useImperativeHandle(ref, callback, deps)
```

I've separately implemented something along the lines of:
```yml
rules:
  customHookDeps:
    - error
    - additionalHooks
      - useEventListener: 1
      - useCustomHook: 0
      - useOtherHook
```
Where the regex syntax can still be supported.

**Comment by quicksnap** [comment]
Food for thought: if ESLint is able to leverage any TypeScript information, there could be a way to type-level annotate hooks accordingly.

**Comment by grncdr** [comment]
I think this discussion would benefit from some clarification of what is possible and what is feasible. To that end, I'm writing below the limits on what I would personally propose. I certainly don't know everything about what can be done with ESLint, so if you read this and think "he doesn't know what he's talking about" please correct me!

### Limits of this proposal

**Couldn't we infer this automatically?**

Not using ESLint. Or alternatively, not without making this ESLint plugin extremely complicated. And even if somebody did that work (and the team was willing to maintain it)

**Could we annotate the "staticness" of a hook in the source code? (using types and/or comments)**

Unfortunately no, the reason is that the ESLint plugin must analyze the locations a variable is _used_ and not where it's declared. At minimum, you would need to annotate a hook every time you import it, since ESLint works on a file-by-file basis.

**Could a type checker do this automatically?**

After reading the above you might think that Typescript or Flow could be leveraged to tell us when a return value is static. After all, they have the global information about values in separate modules that a linter doesn't.

However, neither of them (as far as I'm aware) let you talk about the type of the _implicit environment_ created by a closure. That is, you can't refer to the variables captured by a function. Without this, you can't propagate information about the closed-over variables to the return type. (If the type systems did have this capability, you theoretically wouldn't need to write the dependency arrays at all)

--

**Comment by douglasjunior** [comment]
I think it is possible to pass a parameter to "eslint-plugin-react-hooks" in `.eslintrc`, with the names of the hooks that are static? 

Something like what we do with [globals](https://eslint.org/docs/user-guide/configuring#specifying-globals)?

**Comment by grncdr** [comment]
> I think it is possible to pass a parameter to "eslint-plugin-react-hooks" in `.eslintrc`, with the names of the hooks that are static? 


Yep, that’s what this issue proposes and what I’ve implemented (see my earlier comments for details). I just wanted to clarify that I think the explicit configuration makes the best possible trade off in terms of implementation complexity.

**Comment by quicksnap** [comment]
I think it would be great to have this. Anyone know how to get feedback from a maintainer to see if we can move forward with this?

**Comment by douglasjunior** [comment]
Suggestion: Follow the same idea as the "camelcase" rule and add "allow" option.

https://github.com/eslint/eslint/pull/10783

**Comment by grncdr** [comment]
@douglasjunior could you provide an example of what you mean? I didn’t understand what you wanted to demonstrate with the PR you linked.

**Comment by alirezamirian** [comment]
I'm a little late to the party but I think a better approach would be to infer such cases by tracing the retuned values and see if it's something static. If this is not feasible, or doesn't make sense from a performance point of view, maybe we can at least annotate each custom hook to provide such information in a jsdoc comment block like this:
```
/**
 * Inspired from the format that is suggested in the discussions above:
 * @react-hook: [false, true]
 */
function useToggle(init = false) {
  const [state, setState] = React.useState(init);
  const toggleState = React.useCallback(() => {
    setState(v => !v);
  }, []);
  return [state, toggleState];
}
```

### Advantages of this approach:
- The information about static parts of the return value is encapsulated in the code. So a third-party library can ship this information with the code without needing the user to adjust their es-lint configuration to make it work.
- It's still fairly easy to capture this information from the AST without advanced and heavy AST processing.

### Disadvantages of this approach:
In the mean while that third-party libraries adopt this feature, there is no way to teach eslint-plugin-react-hooks about such static stuff. i.e.

**Comment by grncdr** [comment]
@alirezamirian do you know if ESlint makes it possible/easy to get the AST information for imported modules? I was under the impression it only worked on a single file at a time.

**Comment by alirezamirian** [comment]
@grncdr That's a good point. I'm not an eslint export but I think you are right and we only have Access to Program node corresponding to a single file. The best we can get from the AST in hand is the import statement source. I don't know if there is a utility for resolving AST from an import declaration.

**Comment by kravets-levko** [comment]
bump

**Comment by quicksnap** [comment]
Just another "+1" post, but I'd like to add in that, while there are workarounds, such as using refs or hooks to wrap that kind of logic, it feels unnecessarily boilerplate-y. Having a pragma for ignored values would be so valuable--often devs lazily and dangerously just turn off the whole block and loose safety.

**Comment by Donov4n** [comment]
Bump ...

**Comment by masad-frost** [comment]
Sent a PR that marks hooks as stable https://github.com/facebook/react/pull/20513, though it doesn't do it for destructutred array types which is proposed here.

I think this is a good first step and the team would more likely to accept something simple.

**Comment by grncdr** [comment]
FYI for those still following this issue. I've updated rebased my fork on top of the latest release (`v4.2.0`) and published a new version `@grncdr/eslint-plugin-react-hooks@4.2.0-fix` still works great for me.

**Comment by grncdr** [comment]
> Allowing configuration of the dependency argument position would be useful as well.
> 
> It is currently hard coded to `0` for `additionalHooks`:
> 
> https://github.com/facebook/react/blob/8b580a89d6dbbde8a3ed69475899addef1751116/packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js#L1361
> 
> This allows support for hooks that take more than 2 arguments. Eg.:
> 
> ```js
> useImperativeHandle(ref, callback, deps)
> ```
> 
> I've separately implemented something along the lines of:
> 
> ```yaml
> rules:
>   customHookDeps:
>     - error
>     - additionalHooks
>       - useEventListener: 1
>       - useCustomHook: 0
>       - useOtherHook
> ```
> 
> Where the regex syntax can still be supported.

I actually needed this for the first time yesterday, so I implemented it in my fork. Available now as `@grncdr/eslint-plugin-react-hooks@4.2.0-fix2`

**Comment by adrianhelvik** [comment]
I'd love to be able to add an annotation when importing the hook like this:

```javascript
import /** @react/constant-hook */ useClosure from './useClosure'
```

I would find this pretty useful for implementing stuff dealing with event handling so forth. Defining it when exporting would be even better, but I'd guess that is out of scope for Eslint.

```javascript
import { useCallback, useRef } from 'react'

/**
 * @react/constant-hook
 */
export default function useClosure(fn) {
  const ref = useRef()
  ref.current = fn

  return useCallback((...args) => {
    return ref.current(...args)
  }, [])
}
```

**Comment by grncdr** [comment]
> Defining it when exporting would be even better, but I'd guess that is out of scope for Eslint.

Yes, at least as far as my ESLint skills go.

---

I'm curious, are you saying you are _only_ interested in solutions where you can annotate the static-ness in your source code? (instead of the ESLint config)

If you are ok with defining it in the ESLint config you would just need to switch to my fork and add this to your config:

```json
    "@grncdr/react-hooks/exhaustive-deps": [
      "error",
      {
        "staticHooks": {
          "useClosure": true
        }
      }
```

**Comment by kravets-levko** [comment]
> I'm curious, are you saying you are only interested in solutions where you can annotate the static-ness in your source code? (instead of the ESLint config)

Just my .02: with a config it's very easy to get code and config out of sync, especially if the team is big. Adding annotations to imports is also not great - if you update your hook you'll have to update all imports accordingly. Ideally the hook itself and its annotations should be in the same file (just as JSDocs for example).

P.S. What if I'll have multiple `useClosure` hooks in different folders, and they all have different returning values?

**Comment by grncdr** [comment]
> P.S. What if I'll have multiple `useClosure` hooks in different folders, and they all have different returning values?

This problem already exists with the current eslint plugin:

```jsx
import {useEffect, useState} from 'react'

const useMyState = useState

export function FooComponent() {
  const [x, setX] = useMyState(1)
  useEffect(() => setX(x => x + 1), []) // exhaustive hook deps error, missing dependency `setX`
  return <span>X is {x}</span>
}
```

> Ideally the hook itself and its annotations should be in the same file (just as JSDocs for example).

I agree, but as explained multiple times in this thread this isn't possible as long as ESLint works on one file at a time. As before, I'd love it if somebody showed up and proved me wrong.

**Comment by conradreuter** [comment]
> I agree, but as explained multiple times in this thread this isn't possible as long as ESLint works on one file at a time. As before, I'd love it if somebody showed up and proved me wrong.

For the more real-world cases I can think of, I believe ESLint could attempt to check the import source instead of just the name. I.e. the following should be doable:
```js
import {useState as useWhatever} from 'react'

export function Component() {
  const [x, setX] = useWhatever(1)
  // expecting no error, because it is the second value returned by`useState` imported from `'react'`
  useEffect(() => setX(x => x + 1), [])
  return <span>{x}</span>
}
```

For this to work, the configuration of the ESLint plugin would contain both the filename and the export name.  

If that turns out to be doable, then the following should be doable as well, given an appropriate configuration:
```js
import {useStaticValue} from '../hooks/static-value'

export function Component() {
  const x = useStaticValue()
  // expecting no error, because `useStaticValue` from `path/to/hooks/static-value` is configured to be static
  useEffect(() => doSomething(x), [])
  return <span>{x}</span>
}
```

There will always be edge-cases, yet I think the import-source-and-name approach could be reasonably easy to implement, and cover a lot of real-world use-cases.

**Comment by grncdr** [comment]
@conradreuter that doesn't actually help with the criticisms from @kravets-levko being discussed. Which were:

> Just my .02: with a config it's very easy to get code and config out of sync, especially if the team is big.

and

> What if I'll have multiple useClosure hooks in different folders, and they all have different returning values?

The first criticism is valid: the ergonomics of external configuration _is_ inferior annotations on the **definition** of a hook. The reasons that annotations aren't going to work have (and had already) been covered multiple times in this thread. That's what I was referring to when I said "I'd love it if somebody showed up and proved me wrong".

The second criticism is less valid, because the ESLint plugin is entirely based around local names, and so far this has not been a problem in practice. That was my only reason for bringing up the renaming of hooks: I don't think it's actually a common practice in the real world. I never do it myself, so I'm not going to expend any effort to support it.

A more important point: it's now been 2 years and nobody from the React team has ever commented on or acknowledged this issue. I'll maintain my fork for as long as I'm working on React projects, but I think at this point there's no hope that this will be upstreamed.

**Comment by Donov4n** [comment]
Bump … :)

**Comment by ValentinH** [comment]
I think having such an option would be a great addition to the ecosystem. For example, for Next.js and its `useRouter()` hook. It is not referentially stable but should not be part of the dependency array. In fact, the [Next.js documentation](https://nextjs.org/docs/api-reference/next/router#usage) always omits it from the deps array:
![image](https://user-images.githubusercontent.com/2678610/151445035-0c05aa7b-edf6-425f-93ae-6143110b2977.png)

This brings confusion to a lot of developers as in [this discussion](https://github.com/vercel/next.js/discussions/29403#discussioncomment-1908563).

As a result, this rule is often set as a warning level, like in the official [Next eslint config](https://github.com/vercel/next.js/blob/eea3adc53d217523ca2595b6403f832d060ed2d1/.eslintrc.json#L257). Having the option to list some hooks whose results can/should be omitted from the dependency array would enable to improve these configuration. For example, Next configuration could whitelist by default the `useRouter`.

I've seen so many mistakes being done with effect dependencies by developers who are just used to quickly drop a `// eslint-disable-next-line react-hooks/exhaustive-deps`. I think that being able to whitelist the actual valid use cases would reduce this habit effect and be able to teach "you must follow this rule".

cc @gaearon

**Comment by scottrippey** [comment]
I strongly agree that this config is necessary.  I maintain many libraries with custom hooks, and would absolutely benefit from this feature.  

Just one example: https://github.com/scottrippey/next-current-route has 2 hooks that follow the `useState` pattern:
```js
import { useRouteState, useCurrentRoute } from 'next-current-route';
//...
const [ id, setId ] = useRouteState('id');
//          ^^^^^ stable
const [ query, route ] = useCurrentRoute();
//             ^^^^^ stable
```

**Comment by grncdr** [comment]
> is any chance of implement this? thread active from 2019 and still no solution

@gdbd did you try my fork? It's not an "official" solution but it works. See https://github.com/facebook/react/issues/16873#issuecomment-536346885 for instructions.

**Comment by OrkhanAlikhanov** [comment]
3 years almost but no progress on such a simple but useful addition. For those not wanting to use a fork, we can utilize [patch-package](https://github.com/ds300/patch-package) to maintain our own patch locally. I created [a patch](https://gist.github.com/OrkhanAlikhanov/18ad8988323bb694ea9f6a83348e23d7) out of @grncdr's work in his fork.

**Comment by Tsourdox** [comment]
This would be a great addition for our all the written custom hooks.

**Comment by scottrippey** [comment]
@grncdr Do you have your fork published anywhere?  I'd love to see how you handled it, and it'd be a great to dust it off and get an official PR here.

**Comment by grncdr** [comment]
> @grncdr Do you have your fork published anywhere?  I'd love to see how you handled it, and it'd be a great to dust it off and get an official PR here.

https://github.com/facebook/react/issues/16873#issuecomment-592574971

**Comment by grncdr** [comment]
For those who are interested, I've published a new version `5.0.0-p30d423311.0` that is [rebased on top of the latest upstream changes](https://github.com/facebook/react/compare/main...grncdr:react:eslint-plugin-react-hooks-static-hooks-2022-edition).

The main improvement is fixing compatibility with ESLint 8. Thanks a lot to @luketanner for not only bringing this to my attention, but opening a PR to fix it. 🥇

**Comment by deiga** [comment]
Could someone enlighten me what the downside is of adding a static dependency to the dependency array of a hook? The way I understand it, one can add the static dependency to the array and it shouldn't have any effect and it will appease the linter rule

**Comment by OrkhanAlikhanov** [comment]
@deiga  There isn't a functional difference, but it is also unnecessary and leads to confusion for a moment. The linter allows omitting internally known [list of hooks](https://github.com/facebook/react/blob/f0bba2d4f57ed9b160cada4039e126b05b9e82de/packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js#L154-L163) from  deps array and we would like to be able to configure that list.

**Comment by stuartwmurray** [comment]
+1 in favour of this configuration option, needed for: 
- useImmer from 'use-immer', which is essentially a layer on top of useState
- useNavigate from react-router
- Lots of custom hooks

**Comment by jadshep** [comment]
Bump. This is desperately needed. This issue has been a massive blocker to wide use of custom hooks.

**Comment by zerosheepmoo** [comment]
Any progress?

**Comment by wogns3623** [comment]
> IMHO it would be great if this plugin could detect some common "static" patterns in custom hook, for example if custom hook returns result of `useRef()`/`useCallback(..., [])`/`useMemo(..., [])` etc.

I wrote an eslint plugin based on a forked version of @grncdr with [some ideas](https://github.com/facebook/react/issues/16873#issuecomment-577650731) from @kravets-levko.
https://github.com/wogns3623/eslint-plugin-better-exhaustive-deps

> This is way beyond the scope of what ESLint can do (it would require whole-program taint-tracking) so definitely not going to happen here.

The static value tracking I wrote was based on existing code from the React team, so I don't think it will have the performance issues @grncdr [was worried about](https://github.com/facebook/react/issues/16873#issuecomment-579366832) - feel free to correct me if I'm wrong!

_P.S. English is not my native language, so the sentences may seem strange. Please let me know of any strange things and I'll fix them!_

**Comment by grncdr** [comment]
> I wrote an eslint plugin based on a forked version of @grncdr with [some ideas](https://github.com/facebook/react/issues/16873#issuecomment-577650731) from @kravets-levko. https://github.com/wogns3623/eslint-plugin-better-exhaustive-deps

Nice, and it's probably a good idea to write it as a separate plugin instead of a fork of the facebook one, since it seems there's no interest from the facebook team.

I took a quick look at the repo but since it starts with an initial commit that already has the relevant changes I'm not really sure what you did differently. Do I understand correctly that this can handle situations like the following?

```js
// file: use-toggle.js
import {useCallback, useState} from 'react'

export function useToggle(initialState = false) {
  const [state, setState] = useState(initialState)
  const toggle = useCallback(() => setState(x => !x), [])
  return [state, toggle, useState]
}

// file: some-component.js
import { useCallback } from 'react'
import { useToggle } from './use-toggle.js'

export function SomeComponent() {
  const [enabled, toggleEnabled] = useToggle()
  
  // ignore how contrived this callback is, the important thing is that
  // `toggleEnabled` is detected as static.
  const toggleAndLog = useCallback(() => {
    toggleEnabled()
    
    console.log('toggled state')
    
  }, [])
  
  return <div>
    <div>{enabled ? 'Enabled' : 'Disabled'}</div>
    <button onClick={toggleAndLog}>Toggle</button>
  </div>
}
```

That is, does it propagate "staticness" across files?

---

In any case, I'm hardly doing anything with React these days, so I'd be very happy to pass the torch and recommend your plugin over my forked one.

**Comment by wogns3623** [comment]
> That is, does it propagate "staticness" across files?

No. A static value check option I created, `checkMemoizedVariableIsStatic`, to check if the return value of `useMemo` or `useCallback` is static inside a single component. It still doesn't automatically infer the return value of custom hooks written by users

**Comment by wogns3623** [comment]
For visual explanation, this plugin with `checkMemoizedVariableIsStatic: true` can handle the following cases
```tsx
import { useCallback } from "react";

export function SomeComponent() {
  const [enabled, setEnabled] = useState(false);
  const toggleEnabled = useCallback(() => setEnabled((x) => !x), []);

  // `toggleEnabled` is detected as static.
  const toggleAndLog = useCallback(() => {
    toggleEnabled();

    console.log("toggled state");
  }, []);

  return (
    <div>
      <div>{enabled ? "Enabled" : "Disabled"}</div>
      <button onClick={toggleAndLog}>Toggle</button>
    </div>
  );
}
```

**Comment by grncdr** [comment]
That still seems like a solid improvement. Does it take into account the dependencies array of `useCallback` so that `useCallback(() => {...}, [someProp])` is **not** considered static?

**Comment by wogns3623** [comment]
Yes, of course! If the values in the dependency array are all static, the return value will also be treated as static. If any of the values are non-static, the return value will also be non-static

```tsx
import { useEffect, useCallback, useMemo } from 'react';

export function SomeComponent({ nonStaticValue }: { nonStaticValue: string }) {
  const staticValue = 'some static value';

  // `staticCallback` is treated as static.
  const staticCallback = useCallback(() => {
    console.log(staticValue);
    // if dependency array has static value(even if it is not necessary),
    // return value of useCallback treated as static.
  }, [staticValue]);

  // if any of the dependencies are not static, return value of useCallback treated as non-static.
  const someMemoizedValue = useMemo(() => {
    return staticValue + nonStaticValue;
  }, [nonStaticValue]);

  useEffect(() => {
    console.log(someMemoizedValue);
  }, []);
  // ^^ lint error: React Hook useEffect has a missing dependency: 'someMemoizedValue'.

  useEffect(() => {
    staticCallback();
  }, []); // no lint error

  // ...
}
```

**Comment by huynhducduy** [comment]
Any update on this? because `biome` already implemented that https://github.com/biomejs/biome/pull/1979 . Guess I gotta think about switch to `biome` then :/

**Comment by waynebloss** [comment]
Here was my solution for today - **getting out of the hooks business altogether for things that are "configured once" during app initialization** such as maybe an `axios` instance or whatever. Instead of using a hook to get these things into my components, I'm now planning on using async imports to get them.

I've tested this with my own app and it works great. I think it will be even better than involving hooks for this sort of thing.

The gist of it is this: In my `main` module I import all the things that I need to configure before React even renders and I configure them. Then I do async imports to get my `render` module and my main component. So basically any module other than `main` will just have to do an async import to get the configured thing. **Things imported this way are considered stable by react-hooks/exhaustive-deps.**

<details>
  <summary>Here is some code...</summary>

From the `main.ts` file loaded by Vite from `index.html`

```ts
// Polyfill for the Metadata Reflection API needed by @/platform/container.
import "@abraham/reflection";
// Load our core services... (where things configure themselves using a DI system)
import "@/platform/container";
import "@/platform/log/LoggerService";
import "@/platform/project";
import "@/workbench/container";
import "@/workbench/services/workbench";

const chunks = [
  // The module to call to render from react-dom
  import("@/app/render"),
  // The main React component...
  import("@/workbench"),
];

Promise.all(chunks).then(([{ default: render }, { default: Workbench }]) => {
  render(Workbench);
});
```

Here's a sample from the `render.tsx` file...
```tsx
import React from "react";
import { createRoot } from "react-dom/client";

const container = document.getElementById("my-root");
if (!container) {
  throw new Error("Invalid document. Element with id 'my-root' not found.");
}
const root = createRoot(container);

function render(App: any) {
  const { VITE_STRICTMODE } = import.meta.env;
  const MaybeStrict =
    VITE_STRICTMODE === "false" ? React.Fragment : React.StrictMode;
  root.render(
    <MaybeStrict>
      <StyledEngineProvider injectFirst>
        <ThemeProvider>
          <CssBaseline />
          <App />
        </ThemeProvider>
      </StyledEngineProvider>
    </MaybeStrict>,
  );
}

export default render;
```

And finally, a bit of code that can go into any React component module to get things which were previously configured by `main.tsx` using the DI system:
```tsx
const { Registry } = await import("@/platform/registry");
const { acquireAll } = Registry.of("platform.container");
const { logger, workbench } = acquireAll(["logger", "workbench"]);

// NOTE: The above 3 lines are just a test. I think I can get it down to 1, definitely 2 lines...

export function MyComponent() {
  React.useEffect(() => { logger.log("Loading MyComponent..."); }, []);
  
  // Previously I was doing something like this:
  // const { logger, workbench } = usePlatform("logger", "workbench");
  // but of course, those are not stable values and have to be passed as 
  // dependencies for EVERY hook that uses them!
}
```

_(Yes, I borrowed code structure and ideas from the source of VS Code! The DI system is based on `tsyringe` but without using decorators which I find too troublesome.)_

</details>

**Comment by waynebloss** [comment]
OK so all of that worked great in development, but there is a slight hiccup for production - **Vite won't compile it** because `Top-level await is not available in the configured target environment ("chrome87", "edge88", "es2020", "firefox78", "safari14" + 2 overrides)`

But it's a great idea if I can get it working. At least I have a new problem to move onto instead of looking at this infernal exhaustive-deps policy! :rofl:

**Comment by waynebloss** [comment]
The compile issue was solved in minutes with one simple change to my `vite.config.ts` file:

```ts
  esbuild: {
    supported: {
      "top-level-await": true,
    },
  },
```

**Comment by huynhducduy** [comment]
Bump

**Comment by GregJArnold** [comment]
Bump.  This seems like an obvious feature to implement.

**Comment by allicanseenow** [comment]
With something like `https://usehooks-ts.com/react-hook/use-event-callback` to create a memoized function that tracks the latest values used in the callback while not forcing the component to rerender, you don't even need to use `useCallback` in a lot of cases and being able to omit the function from the dependency array is a big improvement on code readability

**Comment by MaxwellCohen** [comment]
Is this feature still needed, given that we have the React compiler for much smarter memoization of callbacks and values, and useEffectEvent in effects?

**Comment by gynekolog** [comment]
In my opinion, yes. The linter is reporting a value as unstable even though it is actually static. The solution shouldn’t be “there is memoization, so it’s fine.”
